### PR TITLE
Sparse update

### PR DESCRIPTION
**Pull request by Daniel-Dodd**
# Summary:

Minimal working example of sparse variational Gaussian processes. Needs further work before merging to the main package.

Main features are:
 - The addition of sparse variational Gaussian processes
 - Setting trainable parameters via a pytree's
 - Pushing the probabilistic backend to distrax as mentioned in #42
 - Proposal of consistent notation throughout the package that is close to how the maths is written.
 - Minor bug fixes

### gps.py:
-  Added global jitter variable to all GP models.
- Map estimator, static prams resulted in an error → temporally changed this to raise an exception (note this can now be resolved with the recent addition of trainable parameters in future changes).
- Fixed a bug in the NonConjugatePosterior \__repr__ → previously is incorrectly said “Conjugate posterior” when it printed.
- Added \__rmul__ method to the prior → we can now do  both prior * likelihood and likelihood * prior.

### kernels.py:
- Fixed bug in euclidean distance function.
- Added \__add__ and \__mul__ functionality to the kernel class - so that it is closer to the maths e.g. we can type kernel1 + kernel 2 +( kernel 3 * kernel4)
- Revised cross covariance function (it was back to front) → swapped arguments to match how the maths is written, i.e. can now write Ktx = cross_covariance(k, t, x, k.params), whereas before this had to be implemented as Ktx = cross_covariance(k, x, t, k.params).

### parameters.py:
- Noticed bug in parameter transforms that arrises  for kernel combinations e.g. SumKernel(  kernel_set = [kernel1, kernel2]), due to having to apply transforms over list, as well as dictionaries. To fix this, I added a new function 'build_bijectors' that now works recursively over lists and dictionaries to create a pytree of bijectors. Then to obtain constrainers and unconstrainers to recover 'build_transforms ' as before , I apply forward and inverse methods to the pytree of bijectors.
- Fxed the same issue aforementioned above in the 'transform' function.
- Have changed "initialise" functions to add a boolean pytree of same structure as params, 'trainables' that determines which prameters are trainable.
- Added function to stop gradients for non trainable parameters, "stop_grads" that applies jax.lax.stop_gradient to parameters that are set to false in the trainables pytree.

### abstractions.py:
- Revised existing training functions to set trainable parameters.
- Added minimal working data mini-batcher using TensorFlow data.
- Added minimal working mini-batch training function.

## New additions :

These are currently minimal working.

### variational.py:
- Contains variational family class
- Contains variational posterior class, that utilises variational families
- Contains functions to build the ELBO and VFE for a variational posterior object.

### sparse_gps.py:
- Contains minimal working SVGP 

### gauss_hermite_quadrature.py:
- Contains function that does Gauss-Hermite quadrature.

## Notation:
Made notation more intuitive, e.g. Kxf = gram(kernel, test_inputs, x, kernel.params) in current version of GPJax was not intuitive  → it is confusing to have f corresponding to x and, x  corresponding to test_inputs. 

I have provisionally revised the notation:
- x    for the train inputs
- y    for train labels
- t     for test inputs (this might need to be changed to another letter to avoid confusion with time series)

- F     for the latent function modelled as a GP
- Fx    for the latent function, F, at train inputs, x
- Fmu   for the predictive mean of the latent function, F
- Fcov  for the predictive covariance of the latent function, F
- Fvar  for the predictive (diagonal) variance of the latent function, F

- nx    for the number of train inputs, x
- Kxx   for the kernel gram matrix at train inputs, x
- Lx    for lower cholesky decomposition at train inputs, x
- mx    for prior mean at train inputs, x

- nt    for number of test inputs, t
- Ktt   for gram matrix at train inputs, t
- Lt    for lower cholesky decomposition at test inputs, t
- mt    for prior mean at test inputs, t

- Ktx   for cross covariance between test inputs, t, and train inputs, x

and added notation for sparse GPs:

- u for the inducing outputs
- z for inducing inputs
- nz for number of inducing inputs, z
- Kzz for gram matrix at inducing inputs, z
- Lz for lower cholesky decomposition at inducing inputs, z
- Kzx   for cross covariance between test inputs, z, and train inputs, x

and the Gaussian variational family has two parameterations:
- q(u) = N[u; mu,  sqrt . sqrt^T], is the standard (whiten is false).
- q(u) = N[u; Lz . mu + mz,  (Lz . sqrt) . (Lz . sqrt)^T], is the whitened parameterisation.
with variational parameters, mu and sqrt, a vector and lower-triangular matrix respectively to optimise over.

## Other issues
- Variational poster in variational.py is rough. Would be nice if in later version it would accept a GP posterior and variational family as arguments
- In SVGP from sparse_gps.py - it would be good to remove the "q" argument and take in a variational family.
- Would like opinion of how to structure variational class.

### CODE REVIEW

File: gpjax/gps.py [hunk 1]
```diff
@@ -20,6 +18,46 @@
...
+
+DEFAULT_JITTER = get_defaults()["jitter"]
+
+########################################################
+### Gaussian process (GP) notation used in the code: ###
+########################################################
+
+#   - x     for the train inputs
+#   - y     for train labels
+#   - t     for test inputs
```
Comment (thomaspinder) [inline gpjax/gps.py:31]: I don't like using `t` for test points; it feels too ambigous with time indices. Can we use `xtrain` and `xtest`, or `x` and `xstar`?

File: gpjax/gps.py [hunk 2]
```diff
@@ -44,23 +82,27 @@ class Prior(GP):
     kernel: Kernel
     mean_function: tp.Optional[MeanFunction] = Zero()
     name: tp.Optional[str] = "Prior"
+    jitter: tp.Optional[float] = DEFAULT_JITTER
```
Comment (thomaspinder) [inline gpjax/gps.py:85]: I'm not sure that passing the kernel's jittter amount into the GP instantiator is particularly helpful. It may be best left as a global parameter.

File: gpjax/gps.py [hunk 3]
```diff
@@ -106,64 +152,69 @@ class ConjugatePosterior(Posterior):
...
         self,
-        training: Dataset,
+        train_data: Dataset,
         transformations: tp.Dict,
         priors: dict = None,
         static_params: dict = None,
         negative: bool = False,
     ) -> tp.Callable[[Dataset], Array]:
-        x, y = training.X, training.y
+        x, y, nx = train_data.X, train_data.y, train_data.n
```
Comment (thomaspinder) [inline gpjax/gps.py:201]: Can we unpack as `x`, `y` and `n` or `n_data` - `nx` could be confusing as it kinda implies that `x` and `y` might be of different length.

File: gpjax/gps.py [hunk 4]
```diff
@@ -106,64 +152,69 @@ class ConjugatePosterior(Posterior):
...
 
         def mll(
             params: dict,
         ):
             params = transform(params=params, transform_map=transformations)
             if static_params:
-                params = concat_dictionaries(params, transform(static_params))
+                #params = concat_dictionaries(params, transform(static_params))
+                raise NotImplementedError
```
Comment (thomaspinder) [inline gpjax/gps.py:209]: With your `stop_grads` solution, can these lines be removed now?

File: gpjax/kernels.py [hunk 5]
```diff
@@ -43,14 +46,35 @@ def params(self) -> dict:
...
+
+    def _set_kernels(self, kernels: Sequence[Kernel]) -> None:
+        # add kernels to a list, flattening out instances of this class therein
+        kernels_list: List[Kernel] = []
+        for k in kernels:
+            if isinstance(k, self.__class__):
+                kernels_list.extend(k.kernel_set)
+            else:
+                kernels_list.append(k)
+        self.kernel_set = kernels_list
```
Comment (thomaspinder) [inline gpjax/kernels.py:76]: I'm guessing this is based on GPFlow's `Combination` kernel. We should acknowledge this in the docstring.

File: gpjax/likelihoods.py [hunk 6]
```diff
@@ -32,13 +35,12 @@ class Gaussian(Likelihood):
...
         return {"obs_noise": jnp.array([1.0])}
-
+    
     @property
     def link_function(self) -> Callable:
-        def identity_fn(x):
-            return x
-
-        return identity_fn
+        def link_fn(x, params: dict):
```
Comment (thomaspinder) [inline gpjax/likelihoods.py:41]: We should type that `link_fn` returns as Distrax distribution object.

File: gpjax/sparse_gps.py [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+
+        def variance_fn(test_inputs: Array):
+            t = test_inputs
+            Ktt = gram(self.prior.kernel, t, params["kernel"])
+            Kzt = cross_covariance(self.prior.kernel, z, t, params["kernel"])
+            M = solve_triangular(Lz, Kzt, lower=True)
+            Fcov = Ktt - jnp.matmul(M.T, M)
+
+            if not self.q.whiten:
+                M = solve_triangular(Lz.T, M, lower=False)
```
Comment (thomaspinder) [inline gpjax/sparse_gps.py:206]: This conditional might be more cleanly handled as part of the variational family object.

File: gpjax/sparse_gps.py [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+            mz = self.prior.mean(params)(z)
+            mu -= mz
+        
+        def mean_fn(test_inputs: Array):
+            t = test_inputs
+            Kzt = cross_covariance(self.prior.kernel, z, t, params["kernel"])
+            M = solve_triangular(Lz, Kzt, lower=True)
+    
+            if not self.q.whiten:
+                M = solve_triangular(Lz.T, M, lower=False)
```
Comment (thomaspinder) [inline gpjax/sparse_gps.py:179]: This conditional might be more cleanly handled as part of the variational family object.

File: gpjax/sparse_gps.py [hunk 9]
```diff
@@ -0,0 +1,213 @@
...
+        t = test_inputs
+        Ktt = gram(self.prior.kernel, t, params["kernel"])
+        Kzt = cross_covariance(self.prior.kernel, z, t, params["kernel"])
+        M = solve_triangular(Lz, Kzt, lower=True)
+        Fcov = Ktt - jnp.matmul(M.T, M)
+    
+        if not self.q.whiten:
+            M = solve_triangular(Lz.T, M, lower=False)
+            mz = self.prior.mean(params)(z)
+            mu -= mz
```
Comment (thomaspinder) [inline gpjax/sparse_gps.py:151]: This conditional for computing `M` might be more cleanly handled as part of the variational family object.

File: gpjax/variational.py [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+    def elbo(*args, **kwargs):
+        return model.elbo(train_data, transformations)(*args, **kwargs)
+
+    return elbo
+
+
+def VFE(
+    model: VariationalPosterior, train_data: Dataset, transformations: Dict
+) -> Callable[[Array], Array]:
+    @jit
```
Comment (thomaspinder) [inline gpjax/variational.py:97]: Remove

File: requirements.txt [hunk 11]
```diff
@@ -1,6 +1,7 @@
 jax>=0.2.21
 optax>=0.1.0
 chex>=0.0.8
-tensorflow-probability==0.15.0
+distrax>=0.1.2
+tensorflow >= 2.8.0
```
Comment (thomaspinder) [inline requirements.txt:5]: Why do we need this?

**Review by thomaspinder — CHANGES_REQUESTED** [review]
General comments:
- Unit tests are essential for merging this - I can help with this.
- Docstrings on new functionality would be helpful
- In larger functions e.g., SVGP mean and variance fn.s, it'd be nice to have some intermediary comments.

File: gpjax/variational.py [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+from .types import Array, Dataset
+from .config import add_parameter, Identity, Softplus
+from .utils import I, concat_dictionaries
+
+import distrax
+
+Diagonal = distrax.Lambda(forward=lambda x: jnp.diagflat(x), inverse=lambda x: jnp.diagonal(x))
+
+FillDiagonal = distrax.Chain([Diagonal, Softplus])
+FillTriangular = distrax.Chain([tfb.FillTriangular()])
```
Comment (thomaspinder) [inline gpjax/variational.py:18]: Can https://github.com/deepmind/distrax/blob/master/distrax/_src/bijectors/triangular_linear.py be used here?

File: gpjax/variational.py [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+    def __post_init__(self):
+        """Initialize the variational family."""
+        self.num_inducing = self.inducing_inputs.shape[0]
+        add_parameter("inducing_inputs", Identity)
+
+        nz = self.num_inducing
+
+        if self.mu is None:
+            self.mu = jnp.zeros((nz, 1))
+            add_parameter("mu", Identity)
```
Comment (thomaspinder) [inline gpjax/variational.py:47]: Can we give this a more explicit name. `variational_mean` perhaps?

File: gpjax/variational.py [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+
+        if self.sqrt is None:
+            self.sqrt = I(nz)
+            if self.diag:
+                add_parameter("sqrt", FillDiagonal)
+            else:
+                add_parameter("sqrt", FillTriangular)
```
Comment (thomaspinder) [inline gpjax/variational.py:54]: Can we give this a more explicit name. `variational_covariance` perhaps?

File: gpjax/sparse_gps.py [hunk 15]
```diff
@@ -0,0 +1,169 @@
...
+from gpjax.types import Array, Dataset
+from gpjax.utils import I, concat_dictionaries
+
+from .quadrature import gauss_hermite_quadrature
+from .variational import VariationalFamily, VariationalGaussian, VariationalPosterior
+
+DEFAULT_JITTER = get_defaults()["jitter"]
+
+
+def SVGP(posterior: Posterior, variational_family: VariationalFamily) -> VariationalPosterior:
```
Comment (thomaspinder) [inline gpjax/sparse_gps.py:23]: I'm uncomfortable with functions being capitalised - it makes it look like a class. Can we name this something else?

File: gpjax/abstractions.py [hunk 16]
```diff
@@ -52,3 +68,71 @@ def step(params, opt_state):
...
+
+    X, y, n = training.X, training.y, training.n
+    
+    # Make dataloader, set batch size and prefetch buffer:
+    ds = tfd.Dataset.from_tensor_slices((X,y))
+    ds = ds.cache()
+    ds = ds.repeat()
+    ds = ds.shuffle(n)
+    ds = ds.batch(batch_size)
+    ds = ds.prefetch(prefetch_buffer)
```
[the remark this answers is not in the document]
Comment (Daniel-Dodd) [reply gpjax/abstractions.py]: Yes absolutely.

File: gpjax/variational.py (hunk 13 above)
Comment (Daniel-Dodd) [reply gpjax/variational.py]: Addressed in [be2eb5d](https://github.com/thomaspinder/GPJax/pull/43/commits/be2eb5da0b179efba6ddeee4c74ebe4232e56080).

File: gpjax/variational.py (hunk 14 above)
Comment (Daniel-Dodd) [reply gpjax/variational.py]: Addressed in [be2eb5d](https://github.com/thomaspinder/GPJax/pull/43/commits/be2eb5da0b179efba6ddeee4c74ebe4232e56080).

File: gpjax/sparse_gps.py (hunk 15 above)
Comment (Daniel-Dodd) [reply gpjax/sparse_gps.py]: Addressed in [be2eb5d](https://github.com/thomaspinder/GPJax/pull/43/commits/be2eb5da0b179efba6ddeee4c74ebe4232e56080).

File: gpjax/abstractions.py [hunk 17]
```diff
@@ -52,3 +68,71 @@ def step(params, opt_state):
...
+        params = get_params(opt_state)
+        v, g = jax.value_and_grad(loss)(params, batch)
+        return opt_update(i, g, opt_state), v
+    
+    hist = []
+    tr = trange(n_iters) 
+    for i in tr:
+        batch = get_batch()
+        opt_state, v = train_step(i, opt_state, batch)
+        hist.append(v)
```
[the remark this answers is not in the document]
Comment (Daniel-Dodd) [reply gpjax/abstractions.py:154]: Agreed.

File: gpjax/abstractions.py [hunk 18]
```diff
@@ -52,3 +68,71 @@ def step(params, opt_state):
...
+          get_batch,
+          n_iters: tp.Optional[int] = 100,
+          log_rate: tp.Optional[int] = 10,
+          history: tp.Optional[bool] = False
+        )-> tp.Dict:
+    opt_state = opt_init(params)
+    
+    @jax.jit
+    def loss(params, batch):
+        params = stop_grads(params, trainables)
```
[the remark this answers is not in the document]
Comment (Daniel-Dodd) [reply gpjax/abstractions.py:141]: Agreed.

File: gpjax/gps.py (hunk 4 above)
Comment (Daniel-Dodd) [reply gpjax/gps.py]: Yes absolutely.

File: gpjax/kernels.py (hunk 5 above)
Comment (Daniel-Dodd) [reply gpjax/kernels.py:108]: Yes indeed this is.

File: gpjax/abstractions.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (Daniel-Dodd) [reply gpjax/abstractions.py]: This is resolved in [0333b27](https://github.com/thomaspinder/GPJax/pull/43/commits/0333b274c85153788b6d387cd7f00810fc4e5645).

File: gpjax/abstractions.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (Daniel-Dodd) [reply gpjax/abstractions.py]: Resolved in [8c5630f](https://github.com/thomaspinder/GPJax/pull/43/commits/8c5630fab49c889af8b48836eedd30b29e3aef73).

File: gpjax/abstractions.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (thomaspinder) [reply gpjax/abstractions.py]: @Daniel-Dodd to remove lines and push.

File: gpjax/sparse_gps.py (hunk 7 above)
Comment (thomaspinder) [reply gpjax/sparse_gps.py]: @Daniel-Dodd - Lift this code for mean/variance out into the variational family. Have one variational family per whitened/unwhitened representation.

File: gpjax/sparse_gps.py (hunk 8 above)
Comment (thomaspinder) [reply gpjax/sparse_gps.py]: As above

File: gpjax/sparse_gps.py (hunk 9 above)
Comment (thomaspinder) [reply gpjax/sparse_gps.py]: As above

**Comment by thomaspinder** [comment]
@thomaspinder - think about integrating `likelihood(model(x_test))` syntax for predictions. Handles the likelihood noise term in SVGP.

File: gpjax/abstractions.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (Daniel-Dodd) [reply gpjax/abstractions.py]: Resolved in [ca8e33e](https://github.com/thomaspinder/GPJax/pull/43/commits/ca8e33e6ddd2c82c76924610267c749c6be1a8a7).

File: gpjax/kernels.py (hunk 5 above)
Comment (Daniel-Dodd) [reply gpjax/kernels.py]: Resolved in [4c45f51](https://github.com/thomaspinder/GPJax/pull/43/commits/4c45f510b1a685cf611faaba7e67c12329c0de58).
